OsmAnd/src/net/osmand/plus/plugins/odb/VehicleMetricsPlugin.kt

```diff
@@ -0,0 +1,308 @@
+//		return R.drawable.ic_action_car_info
+	}
+
+	override fun getAssetResourceImage(): Drawable? {
+		return app.uiUtilities.getIcon(R.drawable.osmand_development)
+	}
+
+	override fun init(app: OsmandApplication, activity: Activity?): Boolean {
+		for (command in OBDCommand.entries) {
+			OBDDispatcher.addCommand(command)
+		}
+		return true
+	}
+
+	fun registerBooleanPref(prefId: String, defValue: Boolean): CommonPreference<Boolean> {
+		return registerBooleanPreference(prefId, defValue).makeGlobal().makeShared()
+	}
+
+	fun registerIntPref(prefId: String, defValue: Int): CommonPreference<Int> {
+		return registerIntPreference(prefId, defValue).makeGlobal().makeShared()
+	}
+
+	fun registerStringPref(prefId: String, defValue: String?): CommonPreference<String> {
+		return registerStringPreference(prefId, defValue).makeGlobal().makeShared()
+	}
+
+	public override fun registerOptionsMenuItems(
+		mapActivity: MapActivity,
+		helper: ContextMenuAdapter) {
+		if (isActive) {
+			helper.addItem(ContextMenuItem(OsmAndCustomizationConstants.DRAWER_ANT_PLUS_ID)
+				.setTitleId(R.string.obd_plugin_name, mapActivity)
+				.setIcon(R.drawable.ic_action_sensor)
+				.setListener { _: OnDataChangeUiAdapter?, _: View?, _: ContextMenuItem?, _: Boolean ->
+					app.logEvent("obdOpen")
+					OBDMainFragment.showInstance(mapActivity.supportFragmentManager)
+					true
+				})
+		}
+	}
+
+	@SuppressLint("MissingPermission")
+	fun connectToObd(activity: Activity, name: String): Boolean {
+		if (connectedDevice == null) {
+			if (BLEUtils.isBLEEnabled(activity)) {
+				if (AndroidUtils.hasBLEPermission(activity)) {
+					if (socket != null && socket?.isConnected == true) {
+						socket?.close()
+						socket = null
+					}
+					val bluetoothAdapter = BLEUtils.getBluetoothAdapter(activity)
+					bluetoothAdapter?.let { adapter ->
+						LOG.debug("adapter.isDiscovering ${adapter.isDiscovering}")
+						adapter.cancelDiscovery()
+						val pairedDevices = adapter.bondedDevices.toList()
+						pairedDevices.indices
+						pairedDevices.forEach {
+							LOG.debug(it.name)
+						}
+						val obdDevice: BluetoothDevice? =
+							pairedDevices.find { it.name == name }
+						connectedDevice = obdDevice
+						connectToDevice()
+					}
+				} else {
+					AndroidUtils.requestBLEPermissions(activity)
+				}
+			}
+		} else {
+			socket?.close()
+			connectedDevice = null
+			connectToObd(activity, name)
+		}
+		return socket != null && socket?.isConnected == true
+	}
+
+	@SuppressLint("MissingPermission")
+	private fun connectToDevice() {
+		try {
+			socket = connectedDevice?.createRfcommSocketToServiceRecord(uuid)
+			socket?.apply {
+				connect()
+				if (isConnected) {
+					val input = inputStream.source()
+					val output = outputStream.sink()
+					OBDDispatcher.setReadWriteStreams(input, output)
+				}
+			}
+		} catch (error: IOException) {
+			LOG.error("Can't connect to device. $error")
+		}
+	}
+
+	fun addCommandToRead(command: OBDCommand) {
```

Comment:
There should not be references to OBDCommand outside ODB backend. This is completely backend class. There should be some metrics types or classes for operating in UI.